Allow for Homework to be cloned from the web

rposborne opened the pull request:
This module exposes works with a local ruby gem (newline_hw) to allow for rapid cloning and setup of projects.

## Example
![newline_hw_demo_masked](https://cloud.githubusercontent.com/assets/243989/21752414/6c730c10-d5a5-11e6-8d1d-2d021f87cc36.gif)

## Requirements for Review 
- newline_hw 1.1.0 https://github.com/TIYDC/newline_hw
- newline_cli 0.2.5 or higher
- be logged into newline_cli

## Troubleshooting
Tailing the message's passed to `newline_hw`
` tail -f ~/Library/Logs/newline_hw/newlinehw.log`

Get a copy of the logs
`cat ~/Library/Logs/newline_hw/newlinehw.log | pbcopy`

Some notifications to the user depend on https://github.com/theironyard/newline_cli/pull/41 landing and being released. Therefore please only review "authenticated" behavior.

## Known feature adds
- [x] Display link only if the assignment is cloneable.
- [x] Display disabled button when assignment is not cloneable.

rposborne commented:
This is ready for review.

Next steps
Only show clone link when homework is cloneable

rposborne commented:
Ruby gem is pending release so non reviewable atm.

jakerella commented:
Why do you have "help wanted on that labels? What do you need?

rposborne commented on src/scripts/client/modules/hw.js:
```diff
@@ -0,0 +1,119 @@
+      data: {
+        id: submission_id
+      }
+    }, function( response ) {
```
This is solved

rposborne commented on src/scripts/client/modules/hw.js:
```diff
@@ -0,0 +1,119 @@
+      chrome.runtime.sendMessage( {
+        event: "heartbeat",
+        data: {}
+      }, function handleBackgroundPageResponse( res ) {
```
According to https://developer.chrome.com/extensions/runtime#method-sendMessage the callback handler will be called regardless of success or failure.  In the case of failure the `response` is set to null.

jimthedev commented:
@rposborne is it possible to set the directory where this clones to? Also, what happenes if I clone the same hw multiple times (say, because a student updated a hw)? Does the existing folder get overwritten?

rposborne commented:
@jimthedev Yes, the folder is configurable, by default it will `mkdirp ~/theironyard/homework` but it can be anywhere. 

It will blow away the local copy of the submission if it's already present. 

Though all of this behavior is not in the "assistant" but in the gem that it is talking too,  check it out here https://github.com/tiydc/newline_hw.

jakerella commented:
@rposborne Is this ready for me to try to install/run myself from the build/gem? I want to test as a "new user" and see how things go.

rposborne commented:
@jakerella Yes it is, all instructions were just polished.   Build the extension then click the download icon to see install instructions.

jakerella commented:
So... I reloaded the extension, restarted Chrome, updated newline CLI, re-authenticated with newline, installed newline_hw (following all instructions), re-sourced by .bash_profile (no errors), and went to an assignment page. When I click on the download icon in the top right it still says "newline hw not found". I see no relevant errors in the console, although I do see `From Background page Native host has exited.` And of course, the "clone" button does not appear next to any submissions.

Not sure what to do next. Happy to show you @rposborne once you get back to campus.

jakerella commented:
On an unrelated note, one more todo:

- [ ] Add basic instructions to README

jakerella commented on Gruntfile.js:
```diff
@@ -36,11 +36,18 @@ module.exports = function(grunt) {
                     sourceMap: true
                 }
+            },
+            worker: {
```
Do we need this? Should we just use a `copy` task? I think you added it for a possible future worker... but is this premature?

rposborne replied:
It might be premature, but is changing back worth it? Are their drawbacks?

jakerella replied:
Nah, I guess not. Keep it

jakerella commented on src/scripts/worker/worker.js:
```diff
@@ -0,0 +1,101 @@
+(function(tiy){
```
Should we name this worker specifically for the `hw` module? If we add another worker for some other feature we might not want the confusion.

jakerella commented on src/scripts/worker/worker.js:
```diff
@@ -0,0 +1,101 @@
+          itemInQueue.complete = true;
+          itemInQueue.resolve(msg);
+        } else  {
+          itemInQueue.reject(msg);
```
Is it not `complete` in this case? Seems like it would be.

jakerella commented on src/scripts/worker/worker.js:
```diff
@@ -0,0 +1,101 @@
+          itemInQueue.reject(msg);
+        }
+
+        itemInQueue.complete = true;
```
We already did this above... looks like a dup line.

jakerella commented on src/scripts/worker/worker.js:
```diff
@@ -0,0 +1,101 @@
+  let pendingPromises = [];
+
+  tiy.worker.connect = connect;
+  tiy.worker.onNativeMessage = onNativeMessage;
```
Do we need this on the namespace? I don't think we use this in the non-worker JS, do we? Same for the `onDisconnected` handler and `pendingPromises` below...

rposborne replied:
Nope, all exports can be removed.

jakerella commented on src/scripts/worker/worker.js:
```diff
@@ -0,0 +1,101 @@
+    }).catch(function (msg) {
+      sendResponse(msg);
+    });
+    return true;
```
Is this necessary? Seems weird...

rposborne replied:
Yes it is, cause "Google"

https://developer.chrome.com/extensions/messaging#simple

> Note: The sendResponse callback is only valid if used synchronously, or if the event handler returns true to indicate that it will respond asynchronously. The sendMessage function's callback will be invoked automatically if no handlers return true or if the sendResponse callback is garbage-collected.

jakerella commented on src/scripts/client/modules/hw.js:
```diff
@@ -0,0 +1,200 @@
+      $ui.append( `Last Heartbeat at ${moment(resp.message_at).fromNow()}` );
+      $ui.append( `<dl>` );
+
+      for ( let i in resp.data ) {
```
I might rewrite this as:

```
Object.keys(resp.data).forEach(function showData(key) {
  $ui.append( `<dt>${key}</dt><dd>${resp.data[key]}</dd>` );
});
```

(`Object.keys()` won't pull in prototype properties.)

jakerella commented on src/scripts/client/modules/hw.js:
```diff
@@ -0,0 +1,200 @@
+
+
+  function updateUIOnAssignmentPage() {
+    const idFromUrl = uri => Number( uri.substr( uri.lastIndexOf( '/' ) + 1 ) );
```
Hmm... this may be the only fat arrow function in the project. Can we keep things consistent and use `function`?

jakerella replied:
Also... could use a matching regex for this instead?

rposborne replied:
I would hesitate to move to regex, seems like a performance loss.  But I am fine to move off the fat arrow.

jakerella commented on src/scripts/client/modules/hw.js:
```diff
@@ -0,0 +1,200 @@
+
+  function updateUIOnSubmissionPage() {
+    addCloneLinkForSubmissionTo(
+      $( ".edit_assignment_submission" ).first(),
```
Could rewrite to: `$( '.edit_assignment_submission:eq(0)' )`

jakerella commented on src/scripts/client/modules/hw.js:
```diff
@@ -0,0 +1,200 @@
+  function addCloneLinkForSubmissionTo( $el, submission_id, options = {} ) {
+    isSubmissionCloneable( submission_id, function handleIsSubmissionCloneable( resp ) {
+      if ( resp.data.cloneable ) {
+        const clone_link = $el.append(
```
I don't think `.append()` changes the jQuery selection, in which case, you could just keep using `$el` in this function and ditch `clone_link` entirely.

jakerella commented on src/scripts/client/modules/hw.js:
```diff
@@ -0,0 +1,200 @@
+
+        clone_link
+          .find( '.clone_and_open_submission' )
+          .click( function getSubmissionIDandCall( e ) {
```
It would be great to make this a delegated event handler instead (more efficient). Not a blocker, but let me know if you want to try it.

jakerella commented on src/scripts/client/modules/hw.js:
```diff
@@ -0,0 +1,200 @@
+      // Handle native client error message / timeout
+      if ( resp && resp.status === "fail" ) {
+        console.error( resp.message );
+        tiy.showMessage( resp.message );
```
Do we need to `return` here? Or did you want to keep going?

rposborne replied:
I think we want to keep going as this is a utility function it's possible that the caller may want access to the failure state, regardless of the common handling for one. 

Though it was here previously no caller currently uses this behavior.

jakerella commented:
Other than the review comments, two tasks to merge:

- [ ] Add basic instructions to README
- [ ] add docblocks (we have a few, but not many)

jakerella commented on README.md:
```diff
@@ -30,6 +30,23 @@ The gradebook builds on the awesome work of @jwo and @matthiasak. Each student y
+### Homework
+
+The Homework module builds upon the inspiration from @jah2488.  It allows communication to an authorized ruby process running on your computer to rapidly clone and setup git based projects on your computer.  This allows for easy cloning of PR's, Git Repos, Gists, and various other git related links.  For full documentation please read more about the underlying tool newline_hw [here](https://githu ...
+
```
Can we add a bit on how to use it? Basically: install deps, go to assignment/submission page, click link... magic. And add in the animated GIF maybe?

jakerella replied:
I understand not wanting to document in 3 places, but I don't see anywhere where you tell people what to actually do other than install. In other words, if I'm on my path and click on the download icon I do not see a "clone hw" button. I know why - but someone else might not. If you want to document in that dropdown that's cool, but I don't see that. :/

rposborne replied:
Alright, so leave the install instructions, on drop down when no heartbeat.   Show use and upgrade instructions on dropdown when heartbeat detected?

Also add some indicator that they should click back to see if it's installed?

rposborne replied:
When the client is detected
![screenshot 2017-01-10 11 42 53](https://cloud.githubusercontent.com/assets/243989/21815276/0632cd9c-d72a-11e6-8847-993f3176b6a1.png)


When the client is missing
![screenshot 2017-01-10 11 42 35](https://cloud.githubusercontent.com/assets/243989/21815290/0fc70f94-d72a-11e6-9e1a-e16e1cd86754.png)

jakerella replied:
👍

rposborne commented:
Alright, I believe I have addressed the concerns that were raised. 

I have bumped the ruby gem and am passing back the entire config object from `newline_hw` so things like editor etc... will be visible in the UI.  This is to promote knowledge of the settings available and to debug. 
https://github.com/TIYDC/tiyo-assistant/blob/homework/src/scripts/client/modules/hw.js#L55-L69

jakerella commented:
Awesomesauce... let's push this bad boy.

rposborne commented:
:shipit:
